Cookbook/Proposals/MealyReactiveFeedback.md

```diff
@@ -0,0 +1,273 @@
+- Improves [Babylonpartners/ios-playbook#44](https://github.com/Babylonpartners/ios-playbook/pull/44)
+    - Resolves "idle -> routing -> idle" navigation state transition dance.
+- **Backward compatible** with current Moore machine (we can still add as many states for precise control as we want)
+
+### Usage Examples
+
+#### Example 1: Simple (counter, logging, analytics event)
+
+```swift
+let externalEvent = Signal.merge(
+    increment.map { Event.increment },
+    decrement.map { Event.decrement }
+)
+
+let loggingFeedback = Feedback { signal -> Signal<Event> in
+    return signal
+        .on(value: { event, state in // side-effect
+            // NOTE: Now we can also retrieve `event` as well as `state`!
+            print("receiving event = \(event) with state = \(state)")
+        })
+        .ignoreValues() // no feedback loop
+        .promoteValue() // for adjusting type
+}
+
+let analyticsFeedback = Feedback { signal -> Signal<Event> in
+    return signal
+        .filter { event, _ in event == .increment }
+        .on(value: { event, _ in // side-effect
+            // NOTE: Analytics is one of the example that doesn't need state changes at all.
+            analytics.send(event: Analytics.Event.didTapIncrement)
+        })
+        .ignoreValues() // no feedback loop
+        .promoteValue() // for adjusting type
+}
+
+let system = Property(
+    initial: 0,
+    reduce: counterReducer,
+    events: externalEvent,
+    feedbacks: [loggingFeedback, analyticsFeedback]
+)
+```
+
+To simplify the duplicated code above, see `Feedback.noFeedback` in [ios-playbook#102](https://github.com/Babylonpartners/ios-playbook/pull/102).
+
+#### Example 2: Navigation
+
+```swift
+let externalEvent = Signal.merge(
+    didTapA.map { Event.route(Route.pushA) },
+    didTapB.map { Event.route(Route.modalB) }
+)
+
+let navigationFeedback = Feedback { signal -> Signal<Event> in
+    return signal
+        // NOTE: Using `flatMap(.first)` prevents from routing to multiple screens simultaneously
+        .flatMap(.first) { event, _ in
+            guard case let route = event.route else { return .empty }
+            return flowController.handle(route)
+        }
+        .ignoreValues() // no feedback loop
+        .promoteValue() // for adjusting type
+}
+
+let system = Property(
+    initial: 0,
+    reduce: reducer,
+    events: externalEvent,
+    feedbacks: [navigationFeedback]
+)
+```
+
+For navigation feedback rule and `flatMap(.first)`, see also [Babylonpartners/ios-playbook#44](https://github.com/Babylonpartners/ios-playbook/pull/44).
+This is a handy FRP way of managing "isTransitioning" state outside of `State`.
+
+Because this routing is now event-driven rather than state-driven, **we don't need to transit the state back to the original position after completed, which will reduce the intermediate `case`s e.g. `State.isPushingA` and `Event.didFinishPushingA`**.
```

Comment:
will we then still have states like `loading`?

Reply:
I think `loading` is a different case as this very often represents the state actually (the fact that you need to show a loading indicator)

Reply:
In the same manner "showing" something is a state IMO, the difference is that it does not change how the screen looks. State does not necessary needs to map 1-to-1 to specific renderer change.

Reply:
Any states are worth adding if we don't want to replay the same flow with multiple same events given.

Reply:
> Any states are worth adding if we don't want to replay the same flow with multiple same events given.

Exception comes when `Feedback` uses `flatMap(.first)`.
This will mostly behave the same as "if we don't want to replay the same flow with multiple same events given" without adding extra state in `RAF` side.